bss/drivers/l2output/driver.go

```diff
@@ -0,0 +1,230 @@
+	// which we will need to submit a state root. The former is done by simply
+	// adding the submission interval to the latest committed block's timestamp;
+	// the latter inspects the timestamp of the latest block.
+	nextTimestamp, err := d.sroContract.NextTimestamp(callOpts)
+	if err != nil {
+		log.Error(name+" unable to get next block timestamp", "err", err)
+		return nil, nil, err
+	}
+	latestHeader, err := d.cfg.L1Client.HeaderByNumber(ctx, nil)
+	if err != nil {
+		log.Error(name+" unable to retrieve latest header", "err", err)
+		return nil, nil, err
+	}
+	currentTimestamp := big.NewInt(int64(latestHeader.Time))
+
+	// If the submission window has yet to elapsed, we must wait before
+	// submitting our L2 output commitment. Return start as the end value which
+	// will signal that there is no work to be done.
+	if currentTimestamp.Cmp(nextTimestamp) < 0 {
+		log.Info(name+" submission interval has not elapsed",
+			"currentTimestamp", currentTimestamp, "nextTimestamp", nextTimestamp)
+		return start, start, nil
+	}
+
+	log.Info(name+" submission interval has elapsed",
+		"currentTimestamp", currentTimestamp, "nextTimestamp", nextTimestamp)
+
+	// Otherwise the submission interval has elapsed. Transform the next
+	// expected timestamp into its L2 block number, and add one since end is
+	// exclusive.
+	end, err := d.sroContract.ComputeL2BlockNumber(callOpts, nextTimestamp)
+	if err != nil {
+		log.Error(name+" unable to compute next l2 block number", "err", err)
+		return nil, nil, err
+	}
+	end.Add(end, bigOne)
+
+	return start, end, nil
+}
+
+// CraftBatchTx transforms the L2 blocks between start and end into a batch
+// transaction using the given nonce.
+//
+// NOTE: This method SHOULD NOT publish the resulting transaction.
+func (d *Driver) CraftBatchTx(
+	ctx context.Context,
+	start, end, nonce *big.Int,
+) (*types.Transaction, error) {
+
+	name := d.cfg.Name
+
+	log.Info(name+" crafting batch tx", "start", start, "end", end,
+		"nonce", nonce)
+
+	// Fetch the final block in the range, as this is the only state root we
+	// need to submit.
+	nextCheckpointBlock := new(big.Int).Sub(end, bigOne)
+	checkpointBlock, err := d.cfg.L2Client.HeaderByNumber(
+		ctx, nextCheckpointBlock,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch the next expected timestamp that we will submit along with the
+	// state root.
+	callOpts := &bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	}
+	timestamp, err := d.sroContract.NextTimestamp(callOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sanity check that we are submitting against the same expected timestamp.
+	expCheckpointBlock, err := d.sroContract.ComputeL2BlockNumber(
+		callOpts, timestamp,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if nextCheckpointBlock.Cmp(expCheckpointBlock) != 0 {
+		panic(fmt.Sprintf("next expected checkpoint block has changed, "+
+			"want: %d, found: %d", nextCheckpointBlock.Uint64(),
+			expCheckpointBlock.Uint64()))
```

Comment:
I don't think so, if we're executing this logic that means the prior submission happened 25 minutes/100 blocks in the past, which would be a very deep reorg. I changed it to return an error just in case tho, which would be safer.